facebook/OpenBIC
# fby3.5: bb: Support fan control

fby3.5: bb: Support fan control
Summary:
- Support get/set fan control mode.
- Support get/set fan duty.
- Support get fan RPM.

Dependency: #240 

Test plan:
- Build code: Pass
- Fan control: Pass

Log:
1. Get and set FAN control mode.
root@bmc-oob:~# fan-util --auto-mode status
Auto Mode: Normal
root@bmc-oob:~# fan-util --auto-mode disable
root@bmc-oob:~# fan-util --auto-mode status
Auto Mode: Manual
root@bmc-oob:~# fan-util --auto-mode enable
root@bmc-oob:~# fan-util --auto-mode status
Auto Mode: Normal

2. Use manual command in auto mode.
root@bmc-oob:~# fan-util --auto-mode status
Auto Mode: Normal

root@bmc-oob:~# fan-util --get
Fan 0 Speed: 8059 RPM (70%)
Fan 1 Speed: 7324 RPM (70%)
Fan 2 Speed: 8059 RPM (70%)
Fan 3 Speed: 7305 RPM (70%)
Fan 4 Speed: 8070 RPM (70%)
Fan 5 Speed: 7351 RPM (70%)
Fan 6 Speed: 8048 RPM (70%)
Fan 7 Speed: 7324 RPM (70%)
Fan Mode: Normal
FSCD Driver: linear_nic(nic:nic_sensor_temp)
Sensor Fail: None
Fan Fail: None

root@bmc-oob:~# fan-util --set 30
Setting Zone 0 speed to 30%
Setting Zone 1 speed to 30%
Setting Zone 2 speed to 30%
Setting Zone 3 speed to 30%

- Check fan speeds don't change
root@bmc-oob:~# fan-util --get
Fan 0 Speed: 8059 RPM (70%)
Fan 1 Speed: 7315 RPM (70%)
Fan 2 Speed: 8059 RPM (70%)
Fan 3 Speed: 7324 RPM (70%)
Fan 4 Speed: 8070 RPM (70%)
Fan 5 Speed: 7324 RPM (70%)
Fan 6 Speed: 8026 RPM (30%)
Fan 7 Speed: 7315 RPM (70%)
Fan Mode: Normal
FSCD Driver: linear_nic(nic:nic_sensor_temp)
Sensor Fail: None
Fan Fail: None

3. Use manual command in manual mode.
root@bmc-oob:~# fan-util --auto-mode disable
root@bmc-oob:~# fan-util --auto-mode status
Auto Mode: Manual

root@bmc-oob:~# fan-util --get
Fan 0 Speed: 8081 RPM (70%)
Fan 1 Speed: 7324 RPM (70%)
Fan 2 Speed: 8093 RPM (70%)
Fan 3 Speed: 7315 RPM (70%)
Fan 4 Speed: 8115 RPM (70%)
Fan 5 Speed: 7333 RPM (70%)
Fan 6 Speed: 8059 RPM (70%)
Fan 7 Speed: 7333 RPM (70%)
Fan Mode: Manual(No fscd running)
FSCD Driver: Not support in manual mode(No fscd running)
Sensor Fail: Not support in manual mode(No fscd running)
Fan Fail: Not support in manual mode(No fscd running)

- Set fan duty 30
root@bmc-oob:~# fan-util --set 30
Setting Zone 0 speed to 30%
Setting Zone 1 speed to 30%
Setting Zone 2 speed to 30%
Setting Zone 3 speed to 30%
root@bmc-oob:~# fan-util --get
Fan 0 Speed: 2913 RPM (30%)
Fan 1 Speed: 2693 RPM (30%)
Fan 2 Speed: 2834 RPM (30%)
Fan 3 Speed: 2627 RPM (30%)
Fan 4 Speed: 2840 RPM (30%)
Fan 5 Speed: 2677 RPM (30%)
Fan 6 Speed: 2825 RPM (30%)
Fan 7 Speed: 2692 RPM (30%)
Fan Mode: Manual(No fscd running)
FSCD Driver: Not support in manual mode(No fscd running)
Sensor Fail: Not support in manual mode(No fscd running)
Fan Fail: Not support in manual mode(No fscd running)

- Set fan duty 50
root@bmc-oob:~# fan-util --set 50
Setting Zone 0 speed to 50%
Setting Zone 1 speed to 50%
Setting Zone 2 speed to 50%
Setting Zone 3 speed to 50%
root@bmc-oob:~# fan-util --get
Fan 0 Speed: 5683 RPM (50%)
Fan 1 Speed: 5139 RPM (50%)
Fan 2 Speed: 5666 RPM (50%)
Fan 3 Speed: 5148 RPM (50%)
Fan 4 Speed: 5672 RPM (50%)
Fan 5 Speed: 5135 RPM (50%)
Fan 6 Speed: 5655 RPM (50%)
Fan 7 Speed: 5166 RPM (50%)
Fan Mode: Manual(No fscd running)
FSCD Driver: Not support in manual mode(No fscd running)
Sensor Fail: Not support in manual mode(No fscd running)
Fan Fail: Not support in manual mode(No fscd running)

- Set fan duty 100
root@bmc-oob:~# fan-util --set 100
Setting Zone 0 speed to 100%
Setting Zone 1 speed to 100%
Setting Zone 2 speed to 100%
Setting Zone 3 speed to 100%
root@bmc-oob:~# fan-util --get
Fan 0 Speed: 11765 RPM (100%)
Fan 1 Speed: 10731 RPM (100%)
Fan 2 Speed: 11742 RPM (100%)
Fan 3 Speed: 10751 RPM (100%)
Fan 4 Speed: 11695 RPM (100%)
Fan 5 Speed: 10711 RPM (100%)
Fan 6 Speed: 11672 RPM (100%)
Fan 7 Speed: 10770 RPM (100%)
Fan Mode: Manual(No fscd running)
FSCD Driver: Not support in manual mode(No fscd running)
Sensor Fail: Not support in manual mode(No fscd running)
Fan Fail: Not support in manual mode(No fscd running)

4. Use auto command in auto mode.
root@bmc-oob:~# fan-util --auto-mode enable
root@bmc-oob:~# fan-util --auto-mode status
Auto Mode: Normal
root@bmc-oob:~#
root@bmc-oob:~# fan-util --get
Fan 0 Speed: 8138 RPM (70%)
Fan 1 Speed: 7435 RPM (70%)
Fan 2 Speed: 8172 RPM (70%)
Fan 3 Speed: 7416 RPM (70%)
Fan 4 Speed: 8183 RPM (70%)
Fan 5 Speed: 7398 RPM (70%)
Fan 6 Speed: 8115 RPM (70%)
Fan 7 Speed: 7407 RPM (70%)
Fan Mode: Normal
FSCD Driver: linear_nic(nic:nic_sensor_temp)
Sensor Fail: None
Fan Fail: None

5. Error Test.
- Set out of fan index: 0x05
root@bmc-oob:~# bic-util slot1 0xe0 0x02 0x9c 0x9c 0x00 0x10 0xC0 0xF1 0x05 0x00
9C 9C 00 10 31 F1 C9

- Set fan duty more than 100: 0xFF
root@bmc-oob:~# bic-util slot1 0xe0 0x02 0x9c 0x9c 0x00 0x10 0xC0 0xF1 0x02 0xFF
9C 9C 00 10 31 F1 C9
